File: CaCompteKit/Sources/Domain/Rules/GameCatalog.swift
/// Doc 04 « Catalogue et chargement ». Le contenu (définitions + table des moteurs) est fourni
/// par l'appelant — `Domain` ne connaît aucune implémentation concrète de `GameRules` ; c'est
/// `Catalog` qui assemble `GameCatalog.embedded` avec ses JSON et ses moteurs.
public struct GameCatalog: Sendable {
  private let definitionsByID: [String: GameDefinition]
  private let engineTable: [String: @Sendable () -> any GameRules]

  /// Échoue si un JSON référence un moteur absent de `engineTable` — l'exhaustivité voulue
  /// par doc 04 est ainsi vérifiée à la construction, pas découverte au runtime.
  public init(definitions: [GameDefinition], engineTable: [String: @Sendable () -> any GameRules])
    throws
  {
    for definition in definitions where engineTable[definition.engine] == nil {
      throw GameCatalogError.unknownEngine(definition.engine)
    }
    self.definitionsByID = Dictionary(uniqueKeysWithValues: definitions.map { ($0.id, $0) })
    self.engineTable = engineTable
  }

  public func definition(for gameID: String, version: Int) throws -> GameDefinition {
    guard let definition = definitionsByID[gameID], definition.rulesVersion == version else {
      throw GameCatalogError.unknownGame(gameID: gameID, version: version)
    }
    return definition
  }

  public func rules(for gameID: String, version: Int) throws -> any GameRules {
    let definition = try definition(for: gameID, version: version)
    guard let factory = engineTable[definition.engine] else {
      throw GameCatalogError.unknownEngine(definition.engine)
    }
    return factory()
  }

  public var allGames: [GameDefinition] {
    Array(definitionsByID.values)
  }
}

public enum GameCatalogError: Error, Sendable, Equatable {
  case unknownGame(gameID: String, version: Int)
  case unknownEngine(String)
}

File: android/domain/src/main/kotlin/com/quimene/domain/rules/GameCatalog.kt
package com.quimene.domain.rules

/** Miroir de `GameCatalogError.swift`. */
sealed class GameCatalogError(
    message: String,
) : Exception(message) {
    data class UnknownGame(
        val gameID: String,
        val version: Int,
    ) : GameCatalogError("Jeu inconnu : $gameID (version $version)")

    data class UnknownEngine(
        val engineID: String,
    ) : GameCatalogError("Moteur inconnu : $engineID")
}

/**
 * Miroir de `GameCatalog.swift`. `engineTable` associe un `engineID` (`"skyjo.v1"`,
 * `"generic.sum.v1"`, …) à une fabrique de [GameRules] — mêmes clés que le champ `engine` des
 * fichiers `.json` de `spec/games`.
 */
class GameCatalog(
    definitions: List<GameDefinition>,
    private val engineTable: Map<String, () -> GameRules>,
) {
    private val definitionsByID: Map<String, GameDefinition> = definitions.associateBy { it.id }

    init {
        for (definition in definitions) {
            if (definition.engine !in engineTable) {
                throw GameCatalogError.UnknownEngine(definition.engine)
            }
        }
    }

    fun definition(
        gameID: String,
        version: Int,
    ): GameDefinition {
        val definition = definitionsByID[gameID]
        if (definition == null || definition.rulesVersion != version) {
            throw GameCatalogError.UnknownGame(gameID, version)
        }
        return definition
    }

    fun rules(
        gameID: String,
        version: Int,
    ): GameRules {
        val definition = definition(gameID, version)
        val factory = engineTable[definition.engine] ?: throw GameCatalogError.UnknownEngine(definition.engine)
        return factory()
    }

    val allGames: List<GameDefinition> get() = definitionsByID.values.toList()
}
